enclave-runner/src/usercalls/mod.rs

```diff
@@ -330,124 +522,327 @@ impl EnclaveState {
+                                    Err(EnclaveAbort::InvalidUsercall(n))
+                                }
+                                Err(EnclaveAbort::MainReturned) => Err(EnclaveAbort::MainReturned),
+                                Err(EnclaveAbort::Secondary) => Err(EnclaveAbort::Secondary),
+                            };
+                            let _ = tx_clone.try_send((ret, state.mode));
+                        }
+                            .unit_error()
+                            .boxed_local()
+                            .compat();
+                        tokio::runtime::current_thread::spawn(fut);
+                    }
+                    CoResult::Return((tcs, v1, v2)) => {
+                        let fut = async move {
+                            let ret = match state.mode {
+                                EnclaveEntry::Library => {
+                                    enclave_clone.threads_queue.push(StoppedTcs {
+                                        tcs,
+                                        event_queue: state.event_queue,
+                                    });
+                                    Ok((v1, v2))
+                                }
+                                EnclaveEntry::ExecutableMain => Err(EnclaveAbort::MainReturned),
+                                EnclaveEntry::ExecutableNonMain => {
+                                    assert_eq!(
+                                        (v1, v2),
+                                        (0, 0),
+                                        "Expected enclave thread entrypoint to return zero"
+                                    );
+                                    // If the enclave is in the exit-state, threads are no
+                                    // longer able to be launched
+                                    if !enclave_clone.exiting.load(Ordering::SeqCst) {
+                                        enclave_clone.threads_queue.push(StoppedTcs {
+                                            tcs,
+                                            event_queue: state.event_queue,
+                                        });
+                                    }
+                                    Ok((0, 0))
+                                }
+                            };
+                            let _ = tx_clone.try_send((ret, state.mode));
+                        }
+                            .unit_error()
+                            .boxed_local()
+                            .compat();
+                        tokio::runtime::current_thread::spawn(fut);
+                    }
+                };
+            }
+            unreachable!();
+        };
+
+        let select_fut =
+            futures::future::select(return_future.boxed(), io_future.boxed()).map(|either| {
+                match either {
+                    Either::Left((x, _)) => x,
+                    _ => unreachable!(),
```

Comment:
This seems wrong... Why wouldn't `io_future` return here?

Reply:
`io_future` is written in such a way that it never returns. It keeps on looping. 
If I understand the way select works, it means that only the `return_future` can reach this point in the code